tests/test_simplify.py

```diff
@@ -0,0 +1,128 @@
+equation_1 = P(Z1 | (Z2, X)) * P(Z2) * P(Y) * Sum[X](_y * _z3 * _x)
+
+
+class TestMarkovCondition(unittest.TestCase):
+    """Tests for checking the markov condition."""
+
+    def test_markov_raises(self):
+        """Test the type error is raised on invalid input."""
+        for value in [
+            Variable('whatever'),
+            A @ B,
+            A @ ~B,
+            'something else',
+        ]:
+            with self.subTest(value=value), self.assertRaises(TypeError):
+                has_markov_postcondition(value)
+
+    def test_markov_postcondition(self):
+        """Test the expressions have the markov postcondition."""
+        for expression in [
+            P(A),
+            P(A | B),
+            P(A | (B, C)),
+            P(A) * P(B),
+            P(A) * P(A | B),
+            P(A | B) * P(A | C),
+            P(A) / P(B),
+            P(A) / P(A | B),
+            Sum[X](P(A)),
+            Sum[X](P(A | B)),
+            Sum[X](P(A | B) * P(B)),
+        ]:
+            with self.subTest(e=expression):
+                self.assertTrue(has_markov_postcondition(expression))
+
+    def test_missing_markov_postcondition(self):
+        """Test the expressions do not have the markov postcondition."""
+        for expression in [
+            P(A, B),
+            P(A & B | C),
+            P(A & D | (B, C)),
+            P(A, C) * P(B),
+            P(A, C) * P(A | B),
+            P(A) * P(A & C | B),
+            P(A & C | B) * P(A | C),
+            P(A) / P(B & C),
+            P(A & C) / P(B),
+            P(A) / P(A & C | B),
+            Sum[X](P(A, C)),
+            Sum[X](P(A & C | B)),
+            Sum[X](P(A & C | B) * P(B)),
+            Sum[X](P(A | B) * P(B, C)),
+        ]:
+            with self.subTest(e=str(expression)):
+                self.assertFalse(has_markov_postcondition(expression))
+
+
+class TestCanonicalize(unittest.TestCase):
+    """Tests for the canonicalization of a simplified algorithm."""
+
+    def test_canonicalize_raises(self):
+        """Test a value error is raised for non markov-conditioning expressions."""
+        with self.assertRaises(ValueError):
+            canonicalize(P(A, B, C), [A, B, C])
+
+    def test_canonicalize(self):
+        """Test canonicalizing"""
+        for expected, expression, ordering in [
+            (P(A), P(A), [A]),
+            (P(A | B), P(A | B), [A, B]),
+            (P(A | (B, C)), P(A | (B, C)), [A, B, C]),
+            (P(A | (B, C)), P(A | (C, B)), [A, B, C]),
+            (P(A | (B, C, D)), P(A | (B, C, D)), [A, B, C, D]),
+            (P(A | (B, C, D)), P(A | (B, D, C)), [A, B, C, D]),
+            (P(A | (B, C, D)), P(A | (C, B, D)), [A, B, C, D]),
+            (P(A | (B, C, D)), P(A | (C, D, B)), [A, B, C, D]),
+            (P(A | (B, C, D)), P(A | (D, C, B)), [A, B, C, D]),
+            (P(A | (B, C, D)), P(A | (D, B, C)), [A, B, C, D]),
+            (..., P(A) * P(B), [A, B]),
+            (..., P(B) * P(A), [A, B]),
```

Comment:
@djinnome could you provide some more examples? We need a few with sums and fractions